File: src/templates/skills/api-pagination/templates/fastapi/offset_pagination.py
"""
Offset-based Pagination for FastAPI

Simple page-based pagination using offset and limit.
Best for: Small to medium datasets, user-facing pagination with page numbers
"""

from typing import Generic, TypeVar, List
from pydantic import BaseModel, Field
from sqlalchemy.orm import Session
from sqlalchemy import select, func
from math import ceil

T = TypeVar("T")


class OffsetPaginationParams(BaseModel):
    """Query parameters for offset pagination"""
    page: int = Field(default=1, ge=1, description="Page number (1-indexed)")
    limit: int = Field(
        default={{defaultPageSize}},
        ge=1,
        le={{maxPageSize}},
        description="Items per page"
    )

    @property
    def offset(self) -> int:
        """Calculate offset from page number"""
        return (self.page - 1) * self.limit

    class Config:
        extra = "forbid"


class OffsetPaginationMeta(BaseModel):
    """Pagination metadata"""
    current_page: int
    page_size: int
    total_items: int
    total_pages: int
    has_next: bool
    has_previous: bool


class OffsetPaginationResponse(BaseModel, Generic[T]):
    """Response model for offset-based pagination"""
    data: List[T]
    pagination: OffsetPaginationMeta

    class Config:
        from_attributes = True


async def offset_paginate(
    db: Session,
    query: select,
    params: OffsetPaginationParams,
    count_query: Optional[select] = None
) -> tuple[List, int]:
    """
    Perform offset-based pagination on a SQLAlchemy query

    Args:
        db: Database session
        query: SQLAlchemy select query
        params: Pagination parameters
        count_query: Optional separate count query (for optimization)

    Returns:
        Tuple of (items, total_count)

    Example:
        ```python
        from sqlalchemy import select
        from app.models import User

        @app.get("/users", response_model=OffsetPaginationResponse[UserSchema])
        async def list_users(
            page: int = 1,
            limit: int = 20,
            db: Session = Depends(get_db)
        ):
            params = OffsetPaginationParams(page=page, limit=limit)
            query = select(User).order_by(User.created_at.desc())

            items, total = await offset_paginate(db, query, params)

            return OffsetPaginationResponse(
                data=items,
                pagination=OffsetPaginationMeta(
                    current_page=params.page,
                    page_size=params.limit,
                    total_items=total,
                    total_pages=ceil(total / params.limit),
                    has_next=params.page < ceil(total / params.limit),
                    has_previous=params.page > 1
                )
            )
        ```
    """
    # Get total count
    if count_query is None:
        # Use the same query but select count
        count_query = select(func.count()).select_from(query.subquery())

    total_count = db.execute(count_query).scalar() or 0

    # Apply offset and limit
    paginated_query = query.offset(params.offset).limit(params.limit)

    # Execute query
    result = db.execute(paginated_query)
    items = list(result.scalars().all())

    return items, total_count


def create_pagination_response(
    items: List[T],
    total_items: int,
    page: int,
    limit: int
) -> OffsetPaginationResponse[T]:
    """
    Helper function to create pagination response

    Args:
        items: List of items for current page
        total_items: Total number of items
        page: Current page number
        limit: Items per page

    Returns:
        OffsetPaginationResponse with populated metadata
    """
    total_pages = ceil(total_items / limit) if limit > 0 else 0

    return OffsetPaginationResponse(
        data=items,
        pagination=OffsetPaginationMeta(
            current_page=page,
            page_size=limit,
            total_items=total_items,
            total_pages=total_pages,
            has_next=page < total_pages,
            has_previous=page > 1
        )
    )

File: src/templates/skills/api-pagination/templates/express-typescript/offset-pagination.ts
/**
 * Offset-based Pagination Utility for Express + TypeScript
 *
 * Simple page-based pagination using offset and limit.
 * Best for: Small to medium datasets, user-facing pagination with page numbers
 */

import { Request, Response } from 'express';

export interface OffsetPaginationParams {
  page?: number;
  limit?: number;
}

export interface OffsetPaginationResult<T> {
  data: T[];
  pagination: {
    currentPage: number;
    pageSize: number;
    totalItems: number;
    totalPages: number;
    hasNextPage: boolean;
    hasPreviousPage: boolean;
  };
}

export interface OffsetPaginationOptions {
  defaultLimit?: number;
  maxLimit?: number;
}

const DEFAULT_OPTIONS: Required<OffsetPaginationOptions> = {
  defaultLimit: {{defaultPageSize}},
  maxLimit: 100,
};

/**
 * Parse pagination parameters from Express request
 */
export function parseOffsetParams(
  req: Request,
  options: OffsetPaginationOptions = {}
): { offset: number; limit: number; page: number } {
  const opts = { ...DEFAULT_OPTIONS, ...options };

  let page = parseInt(req.query.page as string) || 1;
  let limit = parseInt(req.query.limit as string) || opts.defaultLimit;

  // Validate and constrain
  page = Math.max(1, page);
  limit = Math.min(Math.max(1, limit), opts.maxLimit);

  const offset = (page - 1) * limit;

  return { offset, limit, page };
}

/**
 * Create paginated response
 */
export function createOffsetPaginationResponse<T>(
  data: T[],
  totalItems: number,
  page: number,
  limit: number
): OffsetPaginationResult<T> {
  const totalPages = Math.ceil(totalItems / limit);

  return {
    data,
    pagination: {
      currentPage: page,
      pageSize: limit,
      totalItems,
      totalPages,
      hasNextPage: page < totalPages,
      hasPreviousPage: page > 1,
    },
  };
}

/**
 * Helper to send paginated response
 */
export function sendOffsetPaginationResponse<T>(
  res: Response,
  result: OffsetPaginationResult<T>
): Response {
  return res.json(result);
}

/**
 * Express middleware for offset pagination
 *
 * Usage:
 * ```typescript
 * import { offsetPaginationMiddleware } from './utils/pagination';
 *
 * router.get('/items', offsetPaginationMiddleware(), async (req, res) => {
 *   const { offset, limit, page } = req.pagination;
 *   const items = await db.getItems(offset, limit);
 *   const total = await db.countItems();
 *
 *   return sendOffsetPaginationResponse(
 *     res,
 *     createOffsetPaginationResponse(items, total, page, limit)
 *   );
 * });
 * ```
 */
export function offsetPaginationMiddleware(options: OffsetPaginationOptions = {}) {
  return (req: Request, res: Response, next: Function) => {
    req.pagination = parseOffsetParams(req, options);
    next();
  };
}

// Extend Express Request type
declare global {
  namespace Express {
    interface Request {
      pagination?: { offset: number; limit: number; page: number };
    }
  }
}
